File: src/main/java/io/github/hdzitao/editstarters/utils/JSONConverter.java
package io.github.hdzitao.editstarters.utils;

import com.google.gson.Gson;
import com.intellij.util.xmlb.Converter;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;

/**
 * bean结构持久化
 *
 * @version 3.2.0
 */
public abstract class JSONConverter<T> extends Converter<T> {
    private final Gson gson = new Gson();

    @Override
    public @Nullable T fromString(@NotNull String value) {
        Type type = ((ParameterizedType) getClass().getGenericSuperclass()).getActualTypeArguments()[0];

        return gson.fromJson(value, type);
    }

    @Override
    public @Nullable String toString(@NotNull T value) {
        return gson.toJson(value);
    }
}

File: src/main/kotlin/io/github/hdzitao/editstarters/cache/JSONConverter.kt
package io.github.hdzitao.editstarters.cache

import com.google.gson.Gson
import com.intellij.util.xmlb.Converter
import java.lang.reflect.ParameterizedType

/**
 * bean结构持久化
 *
 * @version 3.2.0
 */
abstract class JSONConverter<T> : Converter<T>() {
    private val gson = Gson()

    override fun fromString(value: String): T {
        val type = (javaClass.genericSuperclass as ParameterizedType).actualTypeArguments[0]

        return gson.fromJson(value, type)
    }

    override fun toString(value: T & Any): String? {
        return gson.toJson(value)
    }
}
